File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        // uisng map 
        
        unordered_map<char,int> mp;
        
        int maxlen = INT_MIN;
        int left=0;
        int right =0;
        
        
        while(right<s.size()){
            char ch = s[right];
            
            if(mp.count(ch)){
                maxlen = max(maxlen,right-left);
                left = max(left,mp[ch]+1);
               
            }
            
            mp[ch]=right;
            right++;
        }
        
        return max(maxlen,right-left);
        
        
        // using set 
        
//         // return the longest substring without repeating chars 
//         set<char> st;
        
//         int l =0;
//         int n = s.size();
//         int ans = 0;
        
//         for(int r=0;r<n;r++){
//             // if s[r] is in set
//             while(st.find(s[r])!=st.end());
//             {
//                 st.erase(s[l]);
//                 l++;
//             }
            
//             st.insert(s[r]);
            
//             ans = max(ans,r-l+1);
//         }
        
//         return ans;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        # give the length of longest substring without repeating characters 

        # with the help of set 

        # return the length right 

        """
        a - 0
        b - 1
        c - 2


        """


        n = len(s)

        mp = collections.defaultdict(int)


        i =0
        j =0
        ans =0

        while j<n:

            # if element is present in set 
            if s[j] in mp:
                i = mp[s[j]]+1




            # if not present in set 
            mp[s[j]] = j

            ans = max(ans,(j-i+1))

            j+=1



        return ans

        



